[EWL-5301] SG2 | Create "Subscribe Promo" Molecule (Cat/Subcat)

csj1012 opened the pull request:
## Ticket(s)

**Jira Ticket**
- [EWL-5301: SG2 | Create "Subscribe Promo" Molecule (Cat/Subcat)](https://issues.ama-assn.org/browse/EWL-5301)

## Description
Adds the "Subscribe Promo" Molecule.


## To Test
- [x] `gulp serve`
- [x] Given that I am viewing SG2
- [x] When I click "Molecules" in the menu
- [x] Then I should see "Subscribe Promo"
- [x] When I click "Subscribe Promo"
- [x] Then I should see the purple box with white text
- [x] And I should optionally see the subject title white all-caps at the top 
- [x] And I should see the the Title in white
- [x] And I should see the subtext in small white text
- [x] And I should see an email input
- [x] And I should see a white "Subscribe" button with purple text
- [x] Did you test in IE 11?

## Visual Regressions

- [x] **Before proceeding:** Run visual regression tests locally to ensure new work does not introduce new bugs.

- [x] _If the new work does your work introduce visual regressions into existing work in the Style Guide (or in a Drupal 8 environment)_, either call these out here or address them before marking the PR as `ready for review`.

- [x] If you are creating or updating a pattern be sure you have created or updated the [scenario in `backstop.json`](ama-style-guide-2/styleguide/backstop.json) and have included a new or updated screenshot in [`bitmaps_reference`](ama-style-guide-2/styleguide/backstop_data/bitmaps_reference).


## Relevant Screenshots/GIFs
<img width="373" alt="screen shot 2018-06-13 at 5 05 18 pm" src="https://user-images.githubusercontent.com/39135832/41381034-f7b741da-6f2b-11e8-931e-382d57cd7aa0.png">


## Remaining Tasks
- [ ] Updating [The Glossary](https://issues.ama-assn.org:8446/confluence/display/DTD/Glossary?src=contextnavpagetreemode) 


## Additional Notes
N/A

---

[Guidelines for Contribution](CONTRIBUTING.md)
[SG2 Standards](ama-style-guide-2/docs/standards.md)

csj1012 committed:
[EWL-5301] Add the atom.

csj1012 committed:
[EWL-5301] Add padding and style the block.

csj1012 committed:
[EWL-5301] Add backstop test and reference images.

csj1012 committed:
[EWL-5301] Add a pattern info file.

labbydev reviewed (changes requested):
Looking at the Zeplin, the "Subscribe" button on the promo should be centered in the container. Also the hover makes the button seemingly disappear. This is something that we should address with UX likely. 

**SG2 Default** 
![screen shot 2018-06-14 at 11 37 15 am](https://user-images.githubusercontent.com/1653723/41426037-13598a1e-6fc8-11e8-82c7-34288c36720c.png)

**SG2 Hover**
![screen shot 2018-06-14 at 11 37 19 am](https://user-images.githubusercontent.com/1653723/41426101-439124a8-6fc8-11e8-8b2e-6efbb1066cc2.png)

----

**Zeplin**
<img width="232" alt="screen shot 2018-06-14 at 11 40 43 am" src="https://user-images.githubusercontent.com/1653723/41426039-137ec73e-6fc8-11e8-905f-2ca65a9a8b23.png">

----
I'm also seeing about 41 backstop failures including:
* atoms -> This isn't failing but it's not testing all of them since the page doesn't scroll correctly. FWIW I don't think we should be doing all atoms together and this test doesn't work anyway since the page doesn't scroll. _If_ we are going to do all atoms together then we need to be using the "open in new window" URL.
* event stub -> change in bottom padding
* hub hero as fifty fifty video -> This is a change in the YouTube preview which I'm willing to look past
* info card as location -> This is a change in the google map which I'm willing to look past.
* search header
* masthead -> in some cases this is a bug in the reference image
* event list -> cascade of changes in event stub molecule
* footer -> change in padding under social media icons
* hub heros -> cascade of hub hero as fifty fifty video
* resource -> change in YouTube screenshot
* event listing -> cascade of changes in event stub molecule

----

Regression in colors

When going to the colors atom in base then I should see swatches of the colors and that functionality is no longer present

![screen shot 2018-06-14 at 11 55 40 am](https://user-images.githubusercontent.com/1653723/41426697-0bc55b32-6fca-11e8-8471-2a338ca5999a.png)

csj1012 commented:
Changes (from UX): 
height of button: 40px / 60px (mobile) 
center button

csj1012 committed:
Merge branch 'develop' into feature/EWL-5301-subscribe-promo

csj1012 committed:
[EWL-5301] Adjust styling for button.

csj1012 commented:
@labbydev I've adjusted the button styling. Recommendation to address the backstop issues and visual regressions separately as they seem to be present in `develop` and outside the scope of this ticket. This is ready for re-review.

JaeBea reviewed (changes requested):
Hi @chelsiejohnston,

I have one update needed to merge and a couple of questions on your implementation.

1. When running backstop I am getting errors for the subscribe promo molecule.
![image](https://user-images.githubusercontent.com/4438120/41557526-945f5614-7303-11e8-94de-5a9098607bcc.png)

Implementation Questions
1. It is my understanding that we should reuse atoms and molecules as much as possible in this style guide. Did UX provide additional information about the html structure that prevents the use of the heading and paragraph atoms?

2. Also some of the added styles seems to repeat the ama__h4 and ama__h2 styles which I believe can be applied to any element. Can those styles be reused here?

csj1012 committed:
[EWL-5301] Replace paragraphs with Twig includes.

csj1012 committed:
[EWL-5301] Update backstop image refs.

csj1012 commented:
Hi @JaeBea - the backstop diff is an intentional change as the button is meant to be centered. I've updated the reference images, sorry about that.

Your understanding about reusing SG patterns is correct. The reason I didn't originally use heading elements was that the JIRA ticket didn't specifically call for them, and that semantically I don't love them being `<h2>`s and `<h4>` because it doesn't really make sense. That said, in doing this I broke the pattern we usually use for the promos (which is to just use the headings anyway). So, I went ahead and just changed them all to the expected pattern includes (which removes the need to extend `ama__h2` and `ama__h4` styles in the named CSS rules).

So anyway, this is ready for re-review! :)

designerneil commented on styleguide/source/assets/scss/02-molecules/_subscribe-promo.scss:
```diff
@@ -0,0 +1,32 @@
+    }
+  }
+
+  button {
```
Shouldn't this be using the button.scss styling?

designerneil replied:
If it's a custom button should we add to the button mixins so we can reuse it?

csj1012 replied:
This styling does already use the button.scss styling but is also different and so needs to override several of the rules there. I don't know if we'll need to reuse it so I'd rather not change this for now. @labbydev @JaeBea do you have any thoughts about this?

labbydev replied:
This seems like a rather niche case, so I'm fine with not adding a button variant for it.

labbydev reviewed (changes requested):
It looks like you forgot to update the hover background color. Other than that everything looks good to me. 

<img width="352" alt="screen shot 2018-06-21 at 11 47 38 am" src="https://user-images.githubusercontent.com/1653723/41733816-b3412960-754a-11e8-9ff7-de80365972d0.png">

csj1012 committed:
[EWL-5301] Update the button hover color per UX convos.

csj1012 committed:
Merge branch 'develop' into feature/EWL-5301-subscribe-promo

csj1012 committed:
[EWL-5301] Update errant backstop failure ref.

csj1012 commented:
All backstop tests are now passing on this branch. I approved and committed one failing test for "visual subcategory exploration with images" that did not appear to have any visual changes inherent. This PR is ready for review.
